include/swift/AST/Pattern.h

```diff
@@ -756,7 +756,93 @@ inline Pattern *Pattern::getSemanticsProvidingPattern() {
     return vp->getSubPattern()->getSemanticsProvidingPattern();
   return this;
 }
-  
+
+/// Describes a pattern and the context in which it occurs.
+class ContextualPattern {
```

Comment:
I wonder if instead of this, a Pattern could instead store a parent which is either a PatternBindingDecl, DeclContext or another Pattern?

Reply:
I thought about that, but didn't want to bloat `Pattern` with a parent pointer. Indeed, I was musing with @CodaFi about removing the `Type` from `Pattern` entirely, because having a `Type` in each pattern is *so* redundant.